### PR TITLE
fix: check if qos0 message check is required

### PR DESCRIPTION
**Pull request by vabhasin**
**Issue #, if available:**
When GG is offline, or when spool storage is full, we check through all messages to see if i we can remove any QOS 0 message to accommodate the incoming message. This loop reads through all messages in spooler and is heavy as it reads through database, in case its a disk spooler. This was discovered as we saw high CPU usage in the Disk spooler performance tests which ran spooler tests with QOS 1 messages and at 100 TPS in an unstable network (online <-> offline every 5 mins)

**Description of changes:**
1. Have a flag to see if we need to check for QOS 0 messages or not, when the removal method is called. We set the flag to true every time a new message is added to spooler queue, and set it to false at the end of removal method. Basically, a flag saying "do we need to check for qos 0 message because the queue has been added to"

**Why is this change necessary:**

**How was this change tested:**
- [x] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 1]
```diff
@@ -250,7 +259,8 @@ public void popOutMessagesWithQosZero() {
 
     private void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
         Iterator<Long> messageIdIterator = queueOfMessageId.iterator();
-        while (messageIdIterator.hasNext() && addJudgementWithCurrentSpoolerSize(needToCheckCurSpoolerSize)) {
+        while (messageIdIterator.hasNext() && addJudgementWithCurrentSpoolerSize(needToCheckCurSpoolerSize)
+                && qos0MessagesInSpooler.get() > 0) {
```
Comment (jcosentino11) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:263]: Super super nit: we can check qos first and avoid function call

**Review by jcosentino11 — DISMISSED** [review]

**Review by MikeDombo — COMMENTED** [review]
Rather than counting, how about a simpler (and less error prone) boolean flag. ex. set to true when _any_ message is added to the queue. If the flag is true, then we will attempt to remove qos 0 messages from the queue when the removal method is called. After calling the removal method, set it back to true. Basically just a flag which means "do we need to check for qos 0 because the queue has been added to"

**Comment by vabhasin** [comment]
> 

Thanks, that's definitely a better way.

**Comment by vabhasin** [comment]
> Rather than counting, how about a simpler (and less error prone) boolean flag. ex. set to true when _any_ message is added to the queue. If the flag is true, then we will attempt to remove qos 0 messages from the queue when the removal method is called. After calling the removal method, set it back to true. Basically just a flag which means "do we need to check for qos 0 because the queue has been added to"


Thanks, that's definitely a better way to do it.

**Review by jcosentino11 — DISMISSED** [review]
Lgtm, unit test?

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 2]
```diff
@@ -39,6 +39,14 @@ public class Spool {
...
     private final AtomicLong nextId = new AtomicLong(0);
     private final BlockingDeque<Long> queueOfMessageId = new LinkedBlockingDeque<>();
+    /**
+     * Flag to see if we need to check for QOS0 messages or not.
+     * It is set to true everytime a new message has been added to spooler queue.
+     * If the flag is true, we will attempt to remove QOS 0 messages from queue
+     * when removeMessagesWithQosZero is called.
+     * Set it back to false at the end of the removeMessagesWithQosZero method.
+     */
+    private boolean qos0MessageCheckRequired = false;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:49]: this must be an atomicboolean

**Comment by vabhasin** [comment]
> Lgtm, unit test?

Existing unit tests should verify that the overall behavior is not changed, testing just this flag part is not possible unless we change a couple of method return types. Went through the logic once again to confirm that when we do set the flag to false, we do it for a valid reason (check removed all qos0 messages or there weren't any in the first place)

**Comment by jcosentino11** [comment]
> > Lgtm, unit test?
> 
> Existing unit tests should verify that the overall behavior is not changed, testing just this flag part is not possible unless we change a couple of method return types. Went through the logic once again to confirm that when we do set the flag to false, we do it for a valid reason:
> 
>     1. Check removed all qos0 messages OR
> 
>     2. There weren't any in the first place OR
> 
>     3. We removed some qos0 messages and stopped midway because size issues were solved, in that case we set it back to true immediately after adding the original message (for which we did this removal process)

Can force disconnect then verify number of times getMessageById is called?

**Review by jcosentino11 — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 3]
```diff
@@ -263,6 +274,7 @@ private void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
                 }
             }
         }
+        qos0MessageCheckRequired = false;
```
[the remark this answers is not in the document]
Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java]: When not removing all messages (For example, to make space for incoming message), it will make sure to set the flag to true after it adds that new message. In case it removes some QOS0 messages and still fails to make space, we are ok with keeping the flag to false (as no need to check for QOS0 messages, as we can't make space even after deleting them)

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 4]
```diff
@@ -168,7 +178,7 @@ public synchronized SpoolMessage addMessage(Publish request) throws InterruptedE
         SpoolMessage message = SpoolMessage.builder().id(id).request(request).build();
         addMessageToSpooler(id, message);
         queueOfMessageId.putLast(id);
-
+        qos0MessageCheckRequired.set(true);
```
Comment (jcosentino11) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:181]: check for qos0 here too?

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:181]: do we also need to set in `addId()` too?

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 5]
```diff
@@ -249,6 +259,9 @@ public void popOutMessagesWithQosZero() {
     }
 
     private void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
+        if (!qos0MessageCheckRequired.get()) {
+            return;
+        }
```
Comment (jcosentino11) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:264]: side note: i think there's a concurrency issue. `removeMessagesWithQosZero` is called 1) when MqttClient connection is interrupted and 2) during spool addMessage.   The latter is synchronized, but the former is not.   Technically a different issue so can be addressed separately, just wanted to call it out

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 4 above)
Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:181]: We don't need the check here, as in this case we want to have it set to true for any new message added. 
Look at case number 3 above in my comment. In that condition, even after doing a check once, it's not guaranteed that there are no more QOS0 messages left in queue, we just stopped because storage issue is solved. Now, _any_ new message addition would change those storage conditions and that's why should set the flag to true again for any new message not just a QOS0 message.

The reason we can have that check in `persistentQueueSync` function is because we don't replace QOS0 messages using `removeMessagesWithQosZero`(The only place where flag could go back to false) while doing the queue sync. 

For the addId(), we don't need it there as well. As that is only called by mqttclient when we try to publish a message (do a `popId` first), but actually fail to publish it and add it to retries (do a `addId`). That's not a new message being added but an old message that was popped from spooler queue (and never actually removed from spooler) but never published and added back to queue.

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 5 above)
Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:264]: Thanks...Nice catch, I can make a note and add it to next PR.

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java]: What about case for mqtt disconnect and reconnect?

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 4 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:181]: I might be missing something, but always setting the flag on addMessage wouldn't give performance benefits?  subsequent calls to addMessage would always perform qos0 checks

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 6]
```diff
@@ -311,6 +325,9 @@ public void persistentQueueSync(Iterable<Long> diskQueueOfIds, CloudMessageSpool
             queueCapacityCheck(request, false);
 
             queueOfMessageId.putLast(currentId);
+            if (request.getQos().equals(QOS.AT_MOST_ONCE)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:328]: switch the order. `QOS.AT_MOST_ONCE.equals(request.getQos())`

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 4 above)
Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:181]: At queue capacity, it will only be triggered at every addMessage if the previous Removal Procedure actually resulted in adding a new message. Otherwise, if it didn’t, the flag will remain false as we exit `addMessage` with SpoolerStorageException. Subsequent addMessage checks are not performed then.

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java (hunk 5 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: still a problem?

Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: Umm..yes, we don't know if some add/remove message could change conditions midway. I'm not sure, will it hurt adding synchronized though?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: If `needToCheckCurSize` is false, maybe just don't check the flag. Always do the cleanup.

Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: We are looping through the whole database (even if we don't have QOS0 messages), if we don't check the flag in that case.

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: Yes, but that's only called on disconnect, not on every single message. So probably not a big deal.

Comment (vabhasin) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: To confirm we are concerned about the additional delay it will cause due to being synchronized right? In that case, wanted to mention...`addMessage` anyways is synchronized (the other place which calls `removeMessagesWithQosZero`).

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: I'm just concerned about a potential deadlock, shouldn't happen, but adding synchronized where it wasn't before does have potential problems.

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: Given that we are using an atomic boolean, this should still be correct

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: If we exit early and don't remove 1 qos 0 message that doesn't seem too bad anyway.

File: src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java [hunk 7]
```diff
@@ -263,8 +263,10 @@ public void popOutMessagesWithQosZero() {
         removeMessagesWithQosZero(false);
     }
 
-    private synchronized void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
-        if (!qos0MessageCheckRequired.get()) {
+    private void removeMessagesWithQosZero(boolean needToCheckCurSpoolerSize) {
+        // Don't proceed if we are doing removal to make space (not due to being offline) and qos0 message check
+        // is not required
+        if (needToCheckCurSpoolerSize && !qos0MessageCheckRequired.get()) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java:269]: Undo this change.

**Review by MikeDombo — APPROVED** [review]

**Review by jcosentino11 — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]
